src/omv/modules/py_fir.c

```diff
@@ -78,352 +145,385 @@ static void fir_fill_image_float_obj(image_t *img, mp_obj_t *data, float min, fl
-            MLX90621_I2CInit(&bus);
-            // The EEPROM must be read at <= 400KHz.
-            cambus_init(&bus, FIR_I2C_ID, CAMBUS_SPEED_FULL);
-
+        case FIR_MLX90621: {
             // parse refresh rate and ADC resolution
-            IR_refresh_rate = py_helper_keyword_int(n_args, args, 1, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_refresh), 64);     // 64Hz
-            ADC_resolution  = py_helper_keyword_int(n_args, args, 2, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_resolution), 18);  // 18-bits
+            int ir_fresh_rate = py_helper_keyword_int(n_args, args, 1, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_refresh), 64);
+            int adc_resolution = py_helper_keyword_int(n_args, args, 2, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_resolution), 18);
 
             // sanitize values
-            ADC_resolution  = ((ADC_resolution > 18) ? 18 : (ADC_resolution < 15) ? 15 : ADC_resolution) - 15;
-            IR_refresh_rate = 14 - __CLZ(__RBIT((IR_refresh_rate > 512) ? 512 : (IR_refresh_rate < 1) ? 1 : IR_refresh_rate));
+            ir_fresh_rate = 14 - __CLZ(__RBIT((ir_fresh_rate > 512) ? 512 : ((ir_fresh_rate < 1) ? 1 : ir_fresh_rate)));
+            adc_resolution = ((adc_resolution > 18) ? 18 : ((adc_resolution < 15) ? 15 : adc_resolution)) - 15;
+
+            fir_mlx_data = xalloc(sizeof(paramsMLX90621));
 
-            mlx_data = xalloc(sizeof(paramsMLX90621));
+            fir_sensor = FIR_MLX90621;
+            FIR_MLX90621_RETRY:
+            cambus_init(&fir_bus, FIR_I2C_ID, CAMBUS_SPEED_FULL); // The EEPROM must be read at <= 400KHz.
+            MLX90621_I2CInit(&fir_bus);
 
             fb_alloc_mark();
-            uint8_t *eeprom = fb_alloc0(256 * sizeof(uint8_t), FB_ALLOC_NO_HINT);
-            int error = 0;
-            error |= MLX90621_DumpEE(eeprom);
+            uint8_t *eeprom = fb_alloc(MLX90621_EEPROM_DATA_SIZE * sizeof(uint8_t), FB_ALLOC_NO_HINT);
+            int error = MLX90621_DumpEE(eeprom);
             error |= MLX90621_Configure(eeprom);
-            error |= MLX90621_SetResolution(ADC_resolution);
-            error |= MLX90621_SetRefreshRate(IR_refresh_rate);
-            error |= MLX90621_ExtractParameters(eeprom, mlx_data);
+            error |= MLX90621_SetRefreshRate(ir_fresh_rate);
+            error |= MLX90621_SetResolution(adc_resolution);
+            error |= MLX90621_ExtractParameters(eeprom, fir_mlx_data);
             fb_alloc_free_till_mark();
 
-            if (error != 0 && first_init == true) {
-                first_init = false;
-                cambus_pulse_scl(&bus);
-                xfree(mlx_data);
-                mlx_data = NULL;
-                goto FIR_MLX90621;
+            if (error != 0) {
+                if (first_init) {
+                    first_init = false;
+                    cambus_pulse_scl(&fir_bus);
+                    goto FIR_MLX90621_RETRY;
+                } else {
+                    nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Failed to init the MLX90621!"));
```

Comment:
```suggestion
                } else {
                    cambus_deinit(&fir_bus);
                    nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Failed to init the MLX90621!"));
```
Deinit the bus here ?

Reply:
The exception causes init0 to be run which deinits the bus. However, should just call py_fir_deinit.

Reply:
Added the deinit function.